lib/vrl/stdlib/src/parse_apache_log.rs

```diff
@@ -0,0 +1,298 @@
+            format,
+            timestamp_format,
+        }))
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "parse apache common log",
+                source: r#"encode_json(parse_apache_log!(s'127.0.0.1 bob frank [10/Oct/2000:13:55:36 -0700] "GET /apache_pb.gif HTTP/1.0" 200 2326', "common"))"#,
+                result: Ok(
+                    indoc! {r#"s'{"host":"127.0.0.1","identity":"bob","message":"GET /apache_pb.gif HTTP/1.0","method":"GET","path":"/apache_pb.gif","protocol":"HTTP/1.0","size":2326,"status":200,"timestamp":"2000-10-10T20:55:36+00:00","user":"frank"}'"#},
+                ),
+            },
+            Example {
+                title: "parse apache combined log",
+                source: r#"encode_json(parse_apache_log!(s'127.0.0.1 bob frank [10/Oct/2000:13:55:36 -0700] "GET /apache_pb.gif HTTP/1.0" 200 2326 "http://www.seniorinfomediaries.com/vertical/channels/front-end/bandwidth" "Mozilla/5.0 (X11; Linux i686; rv:5.0) Gecko/1945-10-12 Firefox/37.0"', "combined"))"#,
+                result: Ok(
+                    indoc! {r#"s'{"agent":"Mozilla/5.0 (X11; Linux i686; rv:5.0) Gecko/1945-10-12 Firefox/37.0","host":"127.0.0.1","identity":"bob","message":"GET /apache_pb.gif HTTP/1.0","method":"GET","path":"/apache_pb.gif","protocol":"HTTP/1.0","referrer":"http://www.seniorinfomediaries.com/vertical/channels/front-end/bandwidth","size":2326,"status":200,"timestamp":"2000-10-10T20:55:36+00:00","user":"frank"}'"#},
+                ),
+            },
+            Example {
+                title: "parse apache error log",
+                source: r#"encode_json(parse_apache_log!(s'[01/Mar/2021:12:00:19 +0000] [ab:alert] [pid 4803:tid 3814] [client 147.159.108.175:24259] I will bypass the haptic COM bandwidth, that should matrix the CSS driver!', "error"))"#,
+                result: Ok(
+                    indoc! {r#"s'{"client":"147.159.108.175","message":"I will bypass the haptic COM bandwidth, that should matrix the CSS driver!","module":"ab","pid":4803,"port":24259,"severity":"alert","thread":"3814","timestamp":"2021-03-01T12:00:19+00:00"}'"#},
+                ),
+            },
+        ]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseApacheLogFn {
+    value: Box<dyn Expression>,
+    format: Bytes,
+    timestamp_format: Option<Box<dyn Expression>>,
+}
+
+impl Expression for ParseApacheLogFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let bytes = self.value.resolve(ctx)?;
+        let message = bytes.try_bytes_utf8_lossy()?;
+        let timestamp_format = match &self.timestamp_format {
+            None => "%d/%b/%Y:%T %z".to_owned(),
+            Some(timestamp_format) => timestamp_format
+                .resolve(ctx)?
+                .try_bytes_utf8_lossy()?
+                .to_string(),
+        };
+
+        let regex = match self.format.as_ref() {
+            b"common" => &*log_util::REGEX_APACHE_COMMON_LOG,
```

Comment:
Wondering why the `&*` is needed here.

Reply:
Since `REGEX_APACHE_COMMON_LOG` is a `lazy_static` it is actually a struct that derefs to the underlying regular expression - hence the `*`. We then want a reference to that regex, so need the `&`.